src/basedtyping/transformer.py

```diff
@@ -0,0 +1,253 @@
+            return self._transform(basedtyping.ForwardRef(value)).body
+        if isinstance(value, int) or (self.string_literals and isinstance(value, str)):
+            return self._literal(node)
+        return node
+
+    @override
+    def visit_Tuple(self, node: ast.Tuple) -> ast.AST:
+        with self.implicit_tuple(value=False):
+            result = self.generic_visit(node)
+        if not self._implicit_tuple:
+            return self.subscript(self._typing("Tuple"), cast(ast.expr, result))
+        return result
+
+    @override
+    def visit_Compare(self, node: ast.Compare) -> ast.AST:
+        if len(node.ops) == 1 and isinstance(node.ops[0], ast.Is):
+            result = self.subscript(
+                self._typing("TypeIs"), cast(ast.expr, self.generic_visit(node.comparators[0]))
+            )
+            return self.generic_visit(result)
+        return self.generic_visit(node)
+
+    @override
+    def visit_IfExp(self, node: ast.IfExp) -> ast.AST:
+        if (
+            isinstance(node.body, ast.Compare)
+            and len(node.body.comparators) == 1
+            and isinstance(node.body.ops[0], ast.Is)
+        ):
+            node.body = self.subscript(
+                self._typing("TypeGuard"),
+                cast(ast.expr, self.generic_visit(node.body.comparators[0])),
+            )
+        return self.generic_visit(node)
+
+    def visit_FunctionType(self, node: ast.FunctionType) -> ast.AST:  # noqa: N802 https://github.com/KotlinIsland/basedmypy/issues/763
+        node = self.generic_visit(node)
+        assert isinstance(node, ast.FunctionType)
+        return ast.Expression(
+            self.subscript(
+                self._typing("Callable"),
+                ast.Tuple([ast.List(node.argtypes, ctx=ast.Load()), node.returns], ctx=ast.Load()),
+            )
+        )
+
+    @override
+    def visit_BinOp(self, node: ast.BinOp) -> ast.AST:
+        node = self.generic_visit(node)
+        if isinstance(node, ast.BinOp) and isinstance(node.op, ast.BitAnd):
+            node = self.subscript(
+                self._basedtyping("Intersection"),
+                ast.Tuple([node.left, node.right], ctx=ast.Load()),
+            )
+        return node
+
+    def _transform(self, value: typing.ForwardRef) -> ast.Expression:
+        tree: ast.AST
+        try:
+            tree = ast.parse(value.__forward_arg__, mode="eval")
+        except SyntaxError:
+            arg = value.__forward_arg__.lstrip()
+            if arg.startswith(("def ", "def(")):
+                arg = arg[3:].lstrip()
+            tree = ast.parse(arg, mode="func_type")
+
+        tree = self.visit(tree)
+        assert isinstance(tree, ast.Expression)
+        return tree
+
+
+def _eval_direct(
+    value: object,
+    globalns: dict[str, object] | None = None,
+    localns: dict[str, object] | None = None,
+) -> object:
+    return eval_type_based(value, globalns, localns, string_literals=False)
```

Comment:
I know it sucks for typing, but `functools.partial` is at least a bit faster than nesting function calls.